contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,445 @@
+        .into_iter()
+        // Can zip without worrying about dropping data as we assert
+        // that lengths are the same above.
+        .zip(token_uris.into_iter())
+        .map(|(token_id, token_uri)| -> StdResult<WasmMsg> {
+            let msg = cw721_base::msg::ExecuteMsg::Mint(cw721_base::MintMsg::<Empty> {
+                token_id,
+                token_uri: Some(token_uri),
+                owner: receiver.to_string(),
+                extension: Empty::default(),
+            });
+            Ok(WasmMsg::Execute {
+                contract_addr: cw721_addr.to_string(),
+                msg: to_binary(&msg)?,
+                funds: vec![],
+            })
+        })
+        .collect::<StdResult<Vec<_>>>()?;
+
+    Ok(Response::default()
+        .add_attribute("method", "execute_mint")
+        .add_messages(mint_messages))
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_do_instantiate_and_mint(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    class_uri: Option<String>,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    if info.sender != env.contract.address {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // Optionally, instantiate a new cw721 contract if one does not
+    // yet exist.
+    let submessages = if CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class_id.clone()) {
+        let expected = CLASS_ID_TO_CLASS_URI.load(deps.storage, class_id.clone())?;
+        if expected != class_uri {
+            // classUri information for a classID shouldn't change
+            // across sends. TODO: is this the case? what are we
+            // suposed to do if not..?
+            return Err(ContractError::ClassUriClash {
+                class_id,
+                expected,
+                actual: class_uri,
+            });
+        }
+        vec![]
+    } else {
+        // Store mapping from classID to classUri. cw721 does not do
+        // this, so we need to do it to stop the infomation from
+        // getting lost.
+        CLASS_ID_TO_CLASS_URI.save(deps.storage, class_id.clone(), &class_uri)?;
+
+        let message = cw721_base::msg::InstantiateMsg {
+            // Name of the collection MUST be class_id as this is how
+            // we create a map entry on reply.
+            name: class_id.clone(),
+            symbol: class_id.clone(), // TODO: What should we put here?
+            minter: env.contract.address.to_string(),
+        };
+        let message = WasmMsg::Instantiate {
+            admin: None, // TODO: Any reason to set ourselves as admin?
+            code_id: CW721_CODE_ID.load(deps.storage)?,
+            msg: to_binary(&message)?,
+            funds: vec![],
+            label: format!("{} ICS721 cw721 backing contract", class_id),
+        };
+        let message = SubMsg::<Empty>::reply_on_success(message, INSTANTIATE_CW721_REPLY_ID);
```

Comment:
Can we collapse this into one line and give the response a more descriptive name? What is the response we expect from the SubMsg call?

Something like:
```suggestion
        let ibc_ack_message = SubMsg::<Empty>::reply_on_success(WasmMsg::Instantiate {
            admin: None, // TODO: Any reason to set ourselves as admin?
            code_id: CW721_CODE_ID.load(deps.storage)?,
            msg: to_binary(&cw721_base::msg::InstantiateMsg {
                // Name of the collection MUST be class_id as this is how
                // we create a map entry on reply.
                name: class_id.clone(),
                symbol: class_id.clone(), // TODO: What should we put here?
                minter: env.contract.address.to_string(),
            })?,
            funds: vec![],
            label: format!("{} ICS721 cw721 backing contract", class_id),
        }, INSTANTIATE_CW721_REPLY_ID);
```